File: Scala/algorithmic_toolbox/src/main/scala-2.12/introduction/SumOfFibonacciNumbers.scala
package introduction

import scala.annotation.tailrec
import utils.RichNumber._

object SumOfFibonacciNumbers {
  private val Modulus: Int = 10

  private def calcPisanoPeriod(modulus: Int): Long = {
    @tailrec
    def loop(a: Int, b: Int, p: Long): Long = {
      if (a == 0 && b == 1) p
      else loop(b, (a + b) % modulus, p + 1)
    }
    loop(a = 1, b = 1, p = 1L)
  }

  private def calcFibonacciModulo(n: Long, modulus: Int): Int =
    (0L until n).foldLeft((0, 1)){ case ((a, b), _) => (b, (a + b) % modulus) }._1

  def calcLastDigitOfTheSumOfFibonacciNumbers(n: Long): Int = {
    val p: Long = calcPisanoPeriod(Modulus)
    (calcFibonacciModulo((n + 2) % p, Modulus) - 1).mod(Modulus)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Long = reader.next().toLong
    val result: Int = calcLastDigitOfTheSumOfFibonacciNumbers(n)
    println(result)
  }
}

File: OCaml/algorithmic_toolbox/introduction/sum_of_fibonacci_numbers.ml
let modulus = 10


let modulo a b =
  let result = a mod b in
    if result >= 0 then result
    else result + b


let calc_pisano_period modulus =
    let rec inner a b p =
        if a = 0 && b = 1 then p
        else inner b ((a + b) mod modulus) (p + 1)
    in inner 1 1 1


let calc_fibonacci_modulo n modulus =
    let rec inner k a b =
        if k < 1 then a
        else inner (k - 1) b ((a + b) mod modulus)
    in inner n 0 1


let calc_last_digit_of_the_sum_of_fibonacci_numbers n =
    let p = calc_pisano_period modulus in
        modulo ((calc_fibonacci_modulo ((n + 2) mod p) modulus) - 1) modulus


let () =
  let n = read_int() in
    print_int (calc_last_digit_of_the_sum_of_fibonacci_numbers n);
    print_newline()
